packages/shopify-api/lib/auth/index.ts

```diff
@@ -1,19 +1,38 @@
 import {ConfigInterface} from '../base-types';
+import {FeatureEnabled, FutureFlagOptions} from '../../future/flags';
 
 import {begin, callback} from './oauth/oauth';
 import {nonce} from './oauth/nonce';
 import {safeCompare} from './oauth/safe-compare';
 import {getEmbeddedAppUrl, buildEmbeddedAppUrl} from './get-embedded-app-url';
+import {TokenExchange, tokenExchange} from './oauth/token-exchange';
 
-export function shopifyAuth(config: ConfigInterface) {
-  return {
+export function shopifyAuth<Config extends ConfigInterface>(
+  config: Config,
+): ShopifyAuth<Config['future']> {
+  const shopify = {
     begin: begin(config),
     callback: callback(config),
     nonce,
     safeCompare,
     getEmbeddedAppUrl: getEmbeddedAppUrl(config),
     buildEmbeddedAppUrl: buildEmbeddedAppUrl(config),
-  };
+  } as ShopifyAuth<Config['future']>;
+
+  if (config.future?.unstable_tokenExchange) {
+    shopify.tokenExchange = tokenExchange(config);
+  }
+
+  return shopify;
 }
 
-export type ShopifyAuth = ReturnType<typeof shopifyAuth>;
+export type ShopifyAuth<Future extends FutureFlagOptions> = {
+  begin: ReturnType<typeof begin>;
+  callback: ReturnType<typeof callback>;
+  nonce: typeof nonce;
+  safeCompare: typeof safeCompare;
+  getEmbeddedAppUrl: ReturnType<typeof getEmbeddedAppUrl>;
+  buildEmbeddedAppUrl: ReturnType<typeof buildEmbeddedAppUrl>;
```

Comment:
These should probably be proper types exported by the files that define the functions as well.